### ISSUE TITLE
Calculate the loss if two or more identical negatives exist in the same batch due to in-batch negatives

### ISSUE THREAD

**Issue by daegonYu** [UKPLab/sentence-transformers#3189]
Hello.

The number of unique corpus is 70,000 data, and the query data is augmented to create 200,000 {question, corpus} pairs. When training with MNR Loss as the loss function and the batch size is about 32k, I think there is a situation where the same negatives can be duplicated due to in-batch negatives. In this case, are the weight updates accumulated as many as the number of duplicated negatives? Are only unique negatives calculated and trained?

**Comment by tomaarsen** [comment]
Hello!

With 70k unique corpus entries across 200k pairs and a batch size of 32k, it's indeed extremely common for this situation:
* anchor_i should have high similarity to positive_i and low similarity to positive_j with j != i, *but* positive_i == positive_j. 

My understanding is that this can inhibit training. Because of this, we added the `no_duplicates` batch sampler:
```python
SentenceTransformersTrainingArguments(
    ...,
    batch_sampler="no_duplicates",
)
```
which avoids *any* duplicate texts in a single batch. This is not ideal, because the batch sampler may be a tad slow and the unused samples are just temporarily skipped for batch X until batch X is "ready", after which those samples are tried again in batch X+1, but it's possible that some samples don't get used at all because of that.

An alternative is avoiding duplicates in the loss itself (to avoid using those as in-batch negatives), but checking for duplicates there is expensive as well, I reckon.

- Tom Aarsen
